torchaudio/models/wav2vec2/model.py

```diff
@@ -116,6 +116,125 @@ def forward(
         return x, lengths
 
 
+class HuBERTModel(Wav2Vec2Model):
```

Comment:
Looking at how `forward` is implemented, I do not see the necessity for this model to inherit `Wav2Vec2Model`.

Also I am wondering if this model should inherit `Wav2Vec2Model` or compose it. https://en.wikipedia.org/wiki/Composition_over_inheritance

I think turning it into composition would allow clearer group of parameter file.
Say we change the model architecture so that an instance of `Wav2Vec2Model` is placed as an attribute of `HuBERTModel`. Then it is easy to extract the related parts of the state dict for fine-tuning.

Reply:
Sounds good. I feel like we should use composition for the existing ``Wav2Vec2Model`` too, by removing ``aux`` layer from it. 
In this way, we have five basic model class: 
```python
Wav2Vec2Model
HuBERTPretrainModel(wav2vec2: Wav2Vec2Model, mask_generator: Module, etc)
Wav2Vec2PretrainModel(wav2vec2: Wav2Vec2Model, extra_training_args)
HuBERTASRModel(wav2vec2: Wav2Vec2Model, aux: Module)
Wav2Vec2ASRModel(wav2vec2: Wav2Vec2Model, aux: Module)
```
The benefit is that when we finish the pre-training of wav2vec2 or HuBERT model, we can easily load the state dictionary by
```python
model = hubert_asr_base()
checkpoint = torch.load(PATH)
model.wav2vec2.load_state_dict(checkpoint["wav2vec2"])
```
I found that in Wav2Vec2Bundle, the model is generated by ``wav2vec2_model``, and ``wav2vec2_base`` or ``huber_base`` is not used in pretrained pipelines. This means we can keep the same API without BC-Breaking. If users want to initialize the model and train from scratch, then they will use ``hubert_pretrain_base`` and ``hubert_asr_base`` because they contain the basic wav2vec2model and other training parameters they need.
What do you think?

Reply:
In ``hubert_pretrain_model`` or ``hubert_asr_model``, we can initialize the model like this:
```python
def hubert_pretrain_model():
    wav2vec2 = wav2vec2_model()
    mask_generator = ...
    final_proj = ...
    return HuBERTPreTrainModel(
        wav2vec2=wav2vec2,
        mask_generator=mask_generator,
        final_proj=final_proj
    )
def hubert_asr_model():
    wav2vec2 = wav2vec2_model()
    aux = ...
    return HuBERTASRModel(
        wav2vec2=wav2vec2,
        aux=aux,
    )
```